text/XXXX-externalize_private_keys.md

```diff
@@ -0,0 +1,111 @@
+[reference-level-explanation]: #reference-level-explanation
+
+Doing the work of generalizing secret storage can be handled by an external package: https://github.com/zalando/go-keyring. go-keyring supports Windos, OS X, and Linux. The key is stored in a `(username,password)` combination, with the "username" being the user's ProfileID (hash of public key), and the password being the base64-encoded private key. OS Keyrings support namespacing, qri's namespace will be a reversed-domain-name constant:
+
+```go
+const qriKeyringServiceID = "io.qri.keys"
+```
+
+Once setup is complete, `config.Profile.ID` stores the ProfileID string this configuration is attached to. Starting a qri process uses this as the "username" in a `Get` call:
+
+```go
+privKey, _ := keyring.Get(qriKeyringServiceID, config.Profile.ID)
+```
+
+### Keys subsystem
+
+All of this work should happen in a new package within the qri codebase: `github.com/qri-io/qri/keys`. We need to takle key rotation next, followed by storing _many_ keys for doing access control. All of this work should happen in a new subsystem called `keys.Service`. `lib.Instance` gets a new field that points to this service.
+
+### Phasing out `config.Profile.PrivKey`
+This work should happen in phases:
+
+1. **copy to keyring, support env var**. In in the next release, add in the new APIs, prefer their use when present.
+2. **intiailize with an empty privKey value**. Once we know the keyring approach works across a number of different platforms in a wide release, switch default initialization behaviour to _not_ write the privKey value.
+3. **remove the field** Eventually, remove the field entirely with a repo migration.
+
+In all phases the privKey field should be the _lowest_ in the loading hierarchy. Keyring values will override `Profile.PrivKey`, and the env var will override all other values.
+
+### Recovering from a deleted `.qri` folder
+Once we've externalized the private keys, we can change the behaviour of `qri setup` to first scan the keyring under the `io.qri.keys` namespace for existing keys. If a key is found, we can ask the user if they'd like to create a new account, or recover an old one.
+
+We can bolster this with calls to the default registry for the discovered public keys to add username info to the list. Profile information may not exist on the default registry, so we'll need to default to showing just the raw ProfileID. We can support this process with docs.
+
+Instead of jumping straight to building the UI for switching, we can supply a `--recover` flag that lets a user supply a ProfileID to recover from:
+
+```
+$ qri setup --recover QmProfileID
+```
+
+listing available private keys in a keyring may also form the basis of profile switching in the future.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### More OS-Specific Work.
+This increases our reliance on an OS-level API that sometimes doesn't exist, and runs the risk of deviating behaviour between platforms. 
+
```

Comment:
Also, Apple could at any point decide that `/usr/bin/security` is a security hole and remove it or remove flags that you need, and require all apps that use keychain to be sandboxed and signed and call the API directly.

Unlikely? Yes.

But you know Apple tightens security all the time. And however unlikely, it would be a huge mess to clean up.

Reply:
This would break many things. Like: all web browsers. Not a reasonable thing to be worried about

Reply:
Agree, probably not reasonable.

But you are missing a bit. They’ll never the API.

go-keyring relies on the cli (`/usr/bin/security`) to avoid calling the API.

I could write a fake VLC.app in an afternoon that prompts for a password to “install” and calls `security` to pull every password out and send them wherever. It _has_ happened in the past.

If it becomes a problem that Apple can find no workaround for, they’d pull the command without blinking.

Reply:
> They’ll never kill the API… go-keyring relies on the cli (`/usr/bin/security`) to avoid calling the API… I could write a fake VLC.app in an afternoon that prompts for a password to “install” and calls `security` to pull every password out and send them wherever. It _has_ happened in the past.

So, just to contribute some Apple-specific context here… the same issue exists with the API. An app is no more restricted from linking the API than it is from spawning another app (e.g. `security`) and both kinds of calls are equally visible to a reviewer if you’re putting the app in the store (or equally invisible if not in the store). If your app is sandboxed, subprocesses you spawn will inherit your entitlements and restrictions. It’s always possible Apple might remove the command-line utility in the future, but it’s extraordinarily unlikely to be because of security concerns about CLI vs. API.